torch/fx/node.py

```diff
@@ -0,0 +1,112 @@
+import dis
+import torch
+import inspect
+import operator
+
+class TraceError(ValueError):
+    pass
+
+# Nodes represent a definition of a value in our graph of operators.
+# For simplicitly, they also have methods defined on them so that they serve as proxy objects for _building_
+# more nodes.
+class Node:
```

Comment:
* How do we handle multiple outputs from the node?

* Where do we store shape/type information for inputs/outputs?
Interested in this from the angle of lowering GraphModule to backends which require known shapes prior compilation/execution.

Reply:
1) Uses of values from a multi-output op will appear via `getitem` calls, for example:
```
import torch

class Foo(torch.nn.Module):
    def forward(self, x):
        topk = torch.topk(x, 3)
        return torch.neg(topk[0]), topk[1]

from torch.fx import symbolic_trace
m = Foo()
gm = symbolic_trace(m)

from tabulate import tabulate
node_specs = [[n.op, n.name, n.target, n.args, n.kwargs] for n in gm.graph.nodes]
print(tabulate(node_specs, headers=['opcode', 'name', 'target', 'args', 'kwargs']))
```

```
opcode         name       target                                                   args         kwargs
-------------  ---------  -------------------------------------------------------  -----------  --------
placeholder    x          x                                                        ()           {}
call_function  topk_1     <built-in method topk of type object at 0x7f74b32d5760>  (x, 3)       {}
call_function  getitem    <built-in function getitem>                              (topk_1, 0)  {}
call_function  neg_1      <built-in method neg of type object at 0x7f74b32d5760>   (getitem,)   {}
call_function  getitem_1  <built-in function getitem>                              (topk_1, 1)  {}
```

2) This is up to the implementation of the analysis that puts shapes on the network, but the simplest shape propagation looks like:

```
class ShapeProp:

    def __init__(self, mod):
        self.root = mod.root
        self.graph = mod.graph
        self.state_dict = self.root.state_dict()
        self.modules = dict(self.root.named_modules())

    def propagate(self, args):
        args_iter = iter(args)
        env = {}

        def load_arg(a):
            return map_arg(a, lambda node: env[node.name])

        for node in self.graph.nodes:
            if node.op == 'placeholder':
                result = next(args_iter)
            elif node.op == 'get_param':
                result = self.state_dict[node.target]
            elif node.op == 'call_function':
                result = node.target(*load_arg(node.args), **load_arg(node.kwargs))
            elif node.op == 'call_method':
                self_obj, *args = load_arg(node.args)
                kwargs = load_arg(node.kwargs)
                result = getattr(self_obj, node.target)(*args, **kwargs)
            elif node.op == 'call_module':
                result = self.modules[node.target](*load_arg(node.args), **load_arg(node.kwargs))

            if isinstance(result, torch.Tensor):
                node.shape = result.shape
                node.dtype = result.dtype

            env[node.name] = result

        return load_arg(self.graph.result)
```

In this case all of the element types/shapes will be stored as `dtype` and `shape` attributes on each `Node` instance. Input shapes can be accessed via the attributes on the corresponding `placeholder` Node, then outputs can be accessed via the `.result` Node on `Graph`